### ISSUE TITLE
Cannot run podman containers as staff_u or user_u under Fedora 34

### ISSUE THREAD

**Issue by laolux** [containers/container-selinux#136]
I have a fresh install of Fedora 34 and cannot run containers with podman as restricted selinux users `staff_u` or `user_u`.
The result is the same for both.
The issue is some transition from `container_runtime_t` to `container_t`.
How to reproduce:
1. Make restricted user via `semanage login -a -s staff_u sestaff`.
2. Login via ssh as `sestaff`
3. Check that `id -Z` shows `staff_u:staff_r:staff_t:s0-s0:c0.c1023`, as expected
4. Run podman container via `podman run --rm -it docker.io/library/alpine:3 sh`; it fails with error code 1 and message `{"msg":"exec container process `/bin/sh`: Permission denied","level":"error","time":"2021-06-28T02:25:38.000627555Z"}
`
5. As root, look at the audit log via `ausearch -m AVC,USER_AVC -ts recent`:
`type=AVC msg=audit(1624847138.627:586): avc:  denied  { transition } for  pid=1229 comm="3" path="/bin/busybox" dev="overlay" ino=12917485 scontext=staff_u:staff_r:container_runtime_t:s0 tcontext=system_u:system_r:container_t:s0:c750,c863 tclass=process permissive=0`
6. Create a new rule to allow the transition with `ausearch -c '3' --raw | audit2allow -M my-3` (content below).
7. Load the rule with `semodule -X 300 -i my-3.pp`
8. Try to run the container again -> it fails with exactly the same AVC issue
9. The container runs fine when `setenforce permissive`, but nobody likes that.

So I am left very confused. First, the container fails to start by default (which it should not). Then, after creating and loading the new rule, it still fails :-(

Here is the `my-3.te` created by `audit2allow`

    module my-3 1.0;

    require {
        type container_runtime_t;
        type container_t;
        class process transition;
    }

    #============= container_runtime_t ==============
    allow container_runtime_t container_t:process transition;

**Comment by rhatdan** [comment]
You are being blocked by a constraint,  Run your AVC through audit2why.

Are you running with the current container-selinux? 2.163.0
rpm -q container-selinux
container-selinux-2.163.0-1.fc34.noarch

**Comment by laolux** [comment]
yes, I am running `container-selinux-2.163.0-1.fc34.noarch` from updates-testing.

audit2why -w returns

     type=AVC msg=audit(1624955388.759:582): avc:  denied  { transition } for  pid=1138 comm="3" path="/bin/busybox" dev="overlay" ino=12917485 scontext=staff_u:staff_r:container_runtime_t:s0 tcontext=system_u:system_r:container_t:s0:c161,c260 tclass=process permissive=0

	Was caused by:
		Missing role allow rule.

		Add an allow rule for the role pair.

enabling dontaudit rules, I get more AVC denials upon running the container:

    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.201:644): avc:  denied  { noatsecure } for  pid=1401 comm="bash" scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:container_runtime_t:s0-s0:c0.c1023 tclass=process permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.202:645): avc:  denied  { rlimitinh } for  pid=1401 comm="podman" scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:container_runtime_t:s0-s0:c0.c1023 tclass=process permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.202:646): avc:  denied  { siginh } for  pid=1401 comm="podman" scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:container_runtime_t:s0-s0:c0.c1023 tclass=process permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.335:647): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.336:648): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.336:649): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.337:650): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=PROCTITLE msg=audit(1624955530.409:651): proctitle="/usr/lib/systemd/systemd-journald"
    type=PATH msg=audit(1624955530.409:651): item=0 name="/proc/1409/exe" inode=29290 dev=00:16 mode=0120777 ouid=1001 ogid=1001 rdev=00:00 obj=staff_u:staff_r:container_runtime_t:s0-s0:c0.c1023 nametype=NORMAL cap_fp=0 cap_fi=0 cap_fe=0 cap_fver=0 cap_frootid=0
    type=CWD msg=audit(1624955530.409:651): cwd="/"
    type=SYSCALL msg=audit(1624955530.409:651): arch=c000003e syscall=267 success=no exit=-13 a0=ffffff9c a1=7ffd7f805a10 a2=557e03e6eb70 a3=1000 items=1 ppid=1 pid=548 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="systemd-journal" exe="/usr/lib/systemd/systemd-journald" subj=system_u:system_r:syslogd_t:s0 key=(null)
    type=AVC msg=audit(1624955530.409:651): avc:  denied  { sys_ptrace } for  pid=548 comm="systemd-journal" capability=19  scontext=system_u:system_r:syslogd_t:s0 tcontext=system_u:system_r:syslogd_t:s0 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.439:652): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.440:653): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.440:654): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.441:655): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.450:656): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.451:657): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.451:658): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.452:659): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.490:660): avc:  denied  { transition } for  pid=1430 comm="3" path="/bin/busybox" dev="overlay" ino=12917485 scontext=staff_u:staff_r:container_runtime_t:s0 tcontext=system_u:system_r:container_t:s0:c408,c948 tclass=process permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.604:661): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:10 2021
    type=AVC msg=audit(1624955530.605:662): avc:  denied  { sys_ptrace } for  pid=762 comm="systemd" capability=19  scontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tcontext=staff_u:staff_r:staff_t:s0-s0:c0.c1023 tclass=cap_userns permissive=0
    ----
    time->Tue Jun 29 17:32:14 2021
    type=AVC msg=audit(1624955534.490:665): avc:  denied  { siginh } for  pid=1453 comm="SetroubleshootP" scontext=system_u:system_r:init_t:s0 tcontext=system_u:system_r:unconfined_service_t:s0 tclass=process permissive=0

I then made rules for all these dontaudit entries. Now the dontaudit stuff does not appear anymore, but the original error is persistent.

**Comment by laolux** [comment]
Ok, updated to `container-selinux-2.164.1-1.git563ba3f.fc34.noarch` from updates-testing. Now running containers as `staff_u` works (thanks a lot), but `user_u` still fails, which same error as before.

**Comment by rhatdan** [comment]
Yes, I have not decided if we want user_u to work.  The issue is user_u could transition to spc_t, which would be a problem. 
What we really need is better SELinux support in Podman to keep the role.

Currently executing Podman is switching from 
staff_u:staff_r:staff_t:s0:c0.c1023@(Podman)container_runtime_exec_t->staff_u:staff_r:container_runtime_t:s0:c0.c1023 -> 
   system_u:system_r:container_t:MCS.

In a better world it would be good to do.
user_u:user_r:user_t:s0:c0.c1023@ (Podman)container_runtime_exec_t->user_u:user_r:user_t:s0:c0.c1023 -> 
   user_u:user_r:container_t:MCS.

Then if the user tried to do 

podman run --security-opt label=type:spc_t ...

SELinux could block, saying 
   user_u:user_r:spc_t:MCS.

Is invalid, since spc_t is not allowed to the user_r role.

**Comment by rhatdan** [comment]
Changes to make this happen would need to be done in github.com://opencontainers/selinux

**Comment by laolux** [comment]
Thanks for the explanations, I think I understand the issue with `user_u` and that a solution might take a while. I was just under the impression that it should work, because according to this [bug report](https://github.com/containers/container-selinux/issues/115) it did work under Fedora 33.
I leave this bug report open so other people running into this issue can find it quickly and know about the current status.

**Comment by bauen1** [comment]
Just noting that I also have an interest in getting this to work (or at least I have some time for testing any potential implementation).

Also related: https://github.com/containers/podman/issues/11185

**Comment by rhatdan** [comment]
Since staff_u is supported.  In order to get user_u support, you would need to eliminate the transition to container_runtime_t and make sure the process could never become spc_t.
Closing, since I think user_u should be handled in separate PR/Issue.
